events.go

```diff
@@ -0,0 +1,290 @@
+package lightstep
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+// Events are emitted by the LightStep tracer as a reporting mechanism. They are
+// handled by adding an OnEvent callback to the Options passed to NewTracer. Events
+// may be cast to specific event types in order access additional information.
+//
+// NOTE: To ensure that events can be accurately identified, each event type contains
+// a sentinel method matching the name of the type. This method is a no-op, it is only used
+// for type coersion.
+type Event interface {
+	Event()
+	String() string
+}
+
+// The ErrorEvent type can be used to filter events. The `Err` method
+// retuns the underlying error.
+type ErrorEvent interface {
+	Event
+	error
+	Err() error
+}
+
+// EventStartError occurs if the Options passed to NewTracer are invalid, and
+// the Tracer has failed to start.
+type EventStartError interface {
+	ErrorEvent
+	EventStartError()
+}
+
+type eventStartError struct {
+	err error
+}
+
+func newEventStartError(err error) *eventStartError {
+	return &eventStartError{err: err}
+}
+
+func (*eventStartError) Event()           {}
+func (*eventStartError) EventStartError() {}
+
+func (e *eventStartError) String() string {
+	return e.err.Error()
+}
+
+func (e *eventStartError) Error() string {
+	return e.err.Error()
+}
+
+func (e *eventStartError) Err() error {
+	return e.err
+}
+
+// EventFlushErrorState lists the possible causes for a flush to fail.
+type EventFlushErrorState string
+
+const (
+	FlushErrorTracerClosed   EventFlushErrorState = "flush failed, the tracer is closed."
+	FlushErrorTracerDisabled EventFlushErrorState = "flush failed, the tracer is disabled."
+	FlushErrorTransport      EventFlushErrorState = "flush failed, could not send report to Collector"
+	FlushErrorReport         EventFlushErrorState = "flush failed, report contained errors"
+)
+
+var (
+	flushErrorTracerClosed   = errors.New(string(FlushErrorTracerClosed))
+	flushErrorTracerDisabled = errors.New(string(FlushErrorTracerDisabled))
```

Comment:
Why not make EventFlushErrorState implement error? Just a 
```
func (s EventFlushErrorState) Error() string { return string(s) }
```
and you can use them as errors, remove these two vars.

Reply:
So, the "implementation of error" happens one level higher, in `EventFlushError`. `EventFlushError` contains two things: the category of error (which are enumerated by `EventFlushErrorState`) and the underlying error. Some error states (`FlushErrorTransport` and  `FlushErrorReport`) could contain a variety of errors that are generated elsewhere: 

```go
  event := newEventFlushError(someNetworkError, FlushErrorTransport)
```

The tracer itself generates the "Flush on Close" and "Flush on Disabled" errors, so those errors happen to live in this package:

```go
  event := newEventFlushError(flushErrorTracerClosed, FlushErrorTracerClosed)
```

The idea was to have flush errors be identifiable on three levels: As an ErrorEvent, a FlushErrorEvent, and type of flush error (EventFlushErrorState). Having each flush error by a separate event type (`EventFlushErrorTracerClosed`, `EventFlushErrorTracerDisabled`, etc) seemed too fine grained... you can't easily identify something as a `FlushError` which is an important distinction. But those type semantics are what I'd like vetted the most... is this sliced up in the most useful way for users to alert and report effectively?

Reply:
👍